### PR TITLE
Add black scholes

### PR DESCRIPTION
**Pull request by dsims21**
## Description
Adds a functions [bsEuropeanOption()] in mathfunctions.java. This function is given by: 

Screen Shot 2023-02-26 at 1 20 56 PM

## Additional context and related issues

These functions are used for options valuation. They are useful for any quantitative finance analyst. I have checked return values and added tests. 

## Release notes

(X ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/operator/scalar/MathFunctions.java [hunk 1]
```diff
@@ -23,15 +23,7 @@
...
 import io.trino.spi.block.Block;
-import io.trino.spi.function.Convention;
-import io.trino.spi.function.Description;
-import io.trino.spi.function.LiteralParameter;
-import io.trino.spi.function.LiteralParameters;
-import io.trino.spi.function.OperatorDependency;
-import io.trino.spi.function.ScalarFunction;
-import io.trino.spi.function.Signature;
-import io.trino.spi.function.SqlNullable;
-import io.trino.spi.function.SqlType;
```
Comment (Praveen2112) [inline core/trino-main/src/main/java/io/trino/operator/scalar/MathFunctions.java:34]: Can we revert these changes.

File: core/trino-main/src/main/java/io/trino/operator/scalar/MathFunctions.java [hunk 2]
```diff
@@ -57,11 +49,7 @@
 import static io.trino.spi.type.Decimals.longTenToNth;
 import static io.trino.spi.type.Decimals.overflows;
 import static io.trino.spi.type.DoubleType.DOUBLE;
-import static io.trino.spi.type.Int128Math.add;
-import static io.trino.spi.type.Int128Math.negate;
-import static io.trino.spi.type.Int128Math.rescale;
-import static io.trino.spi.type.Int128Math.rescaleTruncate;
-import static io.trino.spi.type.Int128Math.subtract;
```
Comment (Praveen2112) [inline core/trino-main/src/main/java/io/trino/operator/scalar/MathFunctions.java:64]: Can we revert these changes.

File: core/trino-main/src/main/java/io/trino/operator/scalar/MathFunctions.java [hunk 3]
```diff
@@ -57,11 +49,7 @@
...
+import static io.trino.spi.type.Int128Math.*;
```
Comment (Praveen2112) [inline core/trino-main/src/main/java/io/trino/operator/scalar/MathFunctions.java:52]: Instead of having import all - can we import the class on-demand ?

File: core/trino-main/src/main/java/io/trino/operator/scalar/MathFunctions.java [hunk 4]
```diff
@@ -1398,4 +1386,30 @@ private static Double mapL2Norm(Block map)
 
         return Math.sqrt(norm);
     }
+
+    @Description("Black-Scholes equation for a Euro vanilla call option")
+    @ScalarFunction("black_scholes_euro_vanilla_call")
+    @SqlType(StandardTypes.DOUBLE)
+    public static double blackScholesEuroVanillaCall(@SqlType(StandardTypes.DOUBLE) double S, @SqlType(StandardTypes.DOUBLE) double K, @SqlType(StandardTypes.DOUBLE) double T, @SqlType(StandardTypes.DOUBLE) double r, @SqlType(StandardTypes.DOUBLE) double sigma) {
```
Comment (Praveen2112) [inline core/trino-main/src/main/java/io/trino/operator/scalar/MathFunctions.java:1393]: Instead of naming argument as S, K, etc, can we have a proper expansion for this. It improves the code readability. Can we place each arguments in new line ? WDYT ?

File: core/trino-main/src/test/java/io/trino/operator/scalar/TestMathFunctions.java [hunk 5]
```diff
@@ -3573,4 +3571,16 @@ public void testWilsonInterval()
         assertThat(assertions.function("wilson_interval_upper", "1250", "1310", "1.96e0"))
                 .isEqualTo(0.9642524717143908);
     }
+
+    @Test
+    public void testBlackScholesEuroVanillaCall() {
```
Comment (Praveen2112) [inline core/trino-main/src/test/java/io/trino/operator/scalar/TestMathFunctions.java:3576]: Can we try with additional edge case values - maybe if one of them could be set as 0.

**Review by Praveen2112 — COMMENTED** [review]
Thanks for working on this PR. We can squash all the 3 commits and can we update the docs for the following function. Please sign the CLA if not signed yet.

File: core/trino-main/src/main/java/io/trino/operator/scalar/MathFunctions.java (hunk 1 above)
Comment (dsims21) [reply core/trino-main/src/main/java/io/trino/operator/scalar/MathFunctions.java]: Reverted

File: core/trino-main/src/main/java/io/trino/operator/scalar/MathFunctions.java (hunk 2 above)
Comment (dsims21) [reply core/trino-main/src/main/java/io/trino/operator/scalar/MathFunctions.java]: Reverted

File: core/trino-main/src/main/java/io/trino/operator/scalar/MathFunctions.java (hunk 3 above)
Comment (dsims21) [reply core/trino-main/src/main/java/io/trino/operator/scalar/MathFunctions.java]: Reverted

File: core/trino-main/src/main/java/io/trino/operator/scalar/MathFunctions.java (hunk 4 above)
Comment (dsims21) [reply core/trino-main/src/main/java/io/trino/operator/scalar/MathFunctions.java]: Renamed args and placed on newline. Please let me know if it looks okay @Praveen2112

File: core/trino-main/src/main/java/io/trino/operator/scalar/MathFunctions.java [hunk 6]
```diff
@@ -1398,4 +1398,36 @@ private static Double mapL2Norm(Block map)
...
+    @Description("Black-Scholes equation for a Euro vanilla option")
+    @ScalarFunction("bs_european_option")
+    @SqlType(StandardTypes.DOUBLE)
+    public static double bsEuropeanOption(
+            @SqlType(StandardTypes.DOUBLE) double stockPrice,
+            @SqlType(StandardTypes.DOUBLE) double strikePrice,
+            @SqlType(StandardTypes.DOUBLE) double timeToExpiration,
+            @SqlType(StandardTypes.DOUBLE) double riskFreeRate,
+            @SqlType(StandardTypes.DOUBLE) double volatility,
+            @SqlType("varchar(x)") String optionType)
```
Comment (dsims21) [inline core/trino-main/src/main/java/io/trino/operator/scalar/MathFunctions.java:1411]: @Praveen2112 can you confirm @SqlType `varchar(x)` and `String` are the best variable types to use here?

File: core/trino-main/src/test/java/io/trino/operator/scalar/TestMathFunctions.java (hunk 5 above)
Comment (dsims21) [reply core/trino-main/src/test/java/io/trino/operator/scalar/TestMathFunctions.java]: I have added more tests which should cover the edge cases.

**Comment by ebyhr** [comment]
@cla-bot check

File: core/trino-main/src/main/java/io/trino/operator/scalar/MathFunctions.java [hunk 7]
```diff
@@ -1398,4 +1398,37 @@ private static Double mapL2Norm(Block map)
...
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/operator/scalar/MathFunctions.java:1411]: I would probably be better to have two functions: `black_scholes_call` and `black_scholes_put`. Simulating an enum with a varchar argument is going to make the computation unnecessarily more expensive.

File: docs/src/main/sphinx/functions/Financial.rst [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    based on the Black-Scholes model.
+
+        SELECT bsEuropeanOption(100, 110, 1, 0.05, 0.2, 'call');
+        -- 5.94327393259265
+        SELECT bsEuropeanOption(100, 90, 0.5, 0.02, 0.3, 'put');
+        -- 5.48771157975784
+
+    ``stockPrice``: A numeric expression representing the current stock price.
+    ``strikePrice``: A numeric expression representing the strike price of the option.
+    ``timeToExpiration``: A numeric expression representing the time in until expiration of the option, in years.
```
Comment (martint) [inline docs/src/main/sphinx/functions/Financial.rst:25]: An interval YEAR-MONTH is a more natural representation for representing durations.

File: core/trino-main/src/main/java/io/trino/operator/scalar/MathFunctions.java [hunk 9]
```diff
@@ -1398,4 +1398,37 @@ private static Double mapL2Norm(Block map)
...
+
+    @Description("Black-Scholes equation for a Euro vanilla option")
+    @ScalarFunction("bs_european_option")
+    @SqlType(StandardTypes.DOUBLE)
+    public static double bsEuropeanOption(
+            @SqlType(StandardTypes.DOUBLE) double stockPrice,
+            @SqlType(StandardTypes.DOUBLE) double strikePrice,
+            @SqlType(StandardTypes.DOUBLE) double timeToExpiration,
+            @SqlType(StandardTypes.DOUBLE) double riskFreeRate,
+            @SqlType(StandardTypes.DOUBLE) double volatility,
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/operator/scalar/MathFunctions.java:1410]: Are there any concerns about the approximate numerics nature of the double type? We should consider using decimal, instead.

**Review by martint — CHANGES_REQUESTED** [review]

**Comment by mosabua** [comment]
:wave: @dsims21 @martint  - this PR has become inactive. We hope you are still interested in working on it. Please let us know, and we can try to get reviewers to help with that.

We're working on closing out old and inactive PRs, so if you're too busy or this has too many merge conflicts to be worth picking back up, we'll be making another pass to close it out in a few weeks.
